pkg/controller/validation/ippool_test.go

```diff
@@ -0,0 +1,644 @@
+				assert.Equal(t, tt.expectedOrigin, result.Origin)
+			}
+		})
+	}
+}
+
+func TestNormalizeCurrentRanges(t *testing.T) {
+	ipRanges := []crdv1beta1.IPRange{
+		{CIDR: "192.168.1.0/24"},
+		{Start: "10.0.0.1", End: "10.0.0.5"},
+		{Start: "2001:db8::1", End: "2001:db8::5"},
+	}
+
+	normalized, err := NormalizeRanges(ipRanges, "")
+	require.NoError(t, err)
+	require.Len(t, normalized, 3)
+
+	assert.Equal(t, "192.168.1.0", normalized[0].Start.String())
+	assert.Equal(t, "192.168.1.255", normalized[0].End.String())
+	assert.Equal(t, "range [192.168.1.0/24]", normalized[0].Origin)
+
+	assert.Equal(t, "10.0.0.1", normalized[1].Start.String())
+	assert.Equal(t, "10.0.0.5", normalized[1].End.String())
+	assert.Equal(t, "range [10.0.0.1-10.0.0.5]", normalized[1].Origin)
+
+	assert.Equal(t, "2001:db8::1", normalized[2].Start.String())
+	assert.Equal(t, "2001:db8::5", normalized[2].End.String())
+	assert.Equal(t, "range [2001:db8::1-2001:db8::5]", normalized[2].Origin)
+}
+
+func TestOverlaps(t *testing.T) {
+	tests := []struct {
+		name     string
+		range1   [2]string
+		range2   [2]string
+		expected bool
+	}{
+		{
+			name:     "no overlap",
+			range1:   [2]string{"192.168.1.1", "192.168.1.10"},
+			range2:   [2]string{"192.168.1.11", "192.168.1.20"},
+			expected: false,
+		},
+		{
+			name:     "partial overlap",
+			range1:   [2]string{"192.168.1.1", "192.168.1.10"},
+			range2:   [2]string{"192.168.1.5", "192.168.1.15"},
+			expected: true,
+		},
+		{
+			name:     "full overlap",
+			range1:   [2]string{"192.168.1.1", "192.168.1.10"},
+			range2:   [2]string{"192.168.1.3", "192.168.1.7"},
+			expected: true,
+		},
+		{
+			name:     "adjacent ranges",
+			range1:   [2]string{"192.168.1.1", "192.168.1.10"},
+			range2:   [2]string{"192.168.1.11", "192.168.1.20"},
+			expected: false,
+		},
+		{
+			name:     "same range",
+			range1:   [2]string{"192.168.1.1", "192.168.1.10"},
+			range2:   [2]string{"192.168.1.1", "192.168.1.10"},
+			expected: true,
+		},
+		{
+			name:     "different IP families",
+			range1:   [2]string{"192.168.1.1", "192.168.1.10"},
+			range2:   [2]string{"2001:db8::1", "2001:db8::a"},
+			expected: false,
+		},
+		{
+			name:     "start equals end of other range",
+			range1:   [2]string{"192.168.1.1", "192.168.1.10"},
+			range2:   [2]string{"192.168.1.10", "192.168.1.20"},
+			expected: true,
+		},
+		{
+			name:     "end equals start of other range",
+			range1:   [2]string{"192.168.1.10", "192.168.1.20"},
+			range2:   [2]string{"192.168.1.1", "192.168.1.10"},
+			expected: true,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			start1, _ := netip.ParseAddr(tt.range1[0])
+			end1, _ := netip.ParseAddr(tt.range1[1])
+			start2, _ := netip.ParseAddr(tt.range2[0])
+			end2, _ := netip.ParseAddr(tt.range2[1])
+
+			result := RangesOverlap(start1, end1, start2, end2)
+			assert.Equal(t, tt.expected, result)
+		})
+	}
+}
+
+func TestValidateIPRangesAndSubnetInfo_EdgeCases(t *testing.T) {
```

Comment:
not clear to me why you don't have a unified `TestValidateIPRangesAndSubnetInfo`?

Reply:
done